dashboard/lib/level_loader.rb

```diff
@@ -1,24 +1,57 @@
 class LevelLoader
+  def self.file_paths(glob)
+    Dir.glob(Rails.root.join(glob)).sort
+  end
+
   def self.for_each_file(path, &block)
-    Dir.glob(Rails.root.join(path)).sort.map(&block)
+    file_paths(path).map(&block)
   end
 
   def self.load_custom_levels
-    level_index = Level.includes(:game).to_a.index_by(&:name)
-    for_each_file('config/scripts/**/*.level') do |path|
-      load_custom_level(path, level_index)
+    # Use a transaction because loading levels requires two separate imports.
+    Level.transaction do
+      existing_level_names = Set.new Level.pluck(:name)
+      level_file_paths = file_paths('config/scripts/**/*.level')
+
+      # First, save stubs of any new levels - they'll need to have ids in
+      # order to create certain associations (in particular
+      # level_concept_difficulty) when we bulk-load the level properties.
+      level_file_names = Set.new level_file_paths.map {|path| name_from_path path}
+      new_level_names = level_file_names - existing_level_names
+      Level.import! new_level_names.map {|name| {name: name}}
+
+      # import! does not give us the ids for the newly created rows, so load
+      # all level models for subsequent updates.
+      level_index = Level.all.index_by(&:name)
+
+      # Load level properties from disk and build a collection of levels that
+      # have changed.
+      changed_levels = level_file_paths.inject([]) do |changed, path|
+        level = load_custom_level(path, level_index)
+        changed << level if level.changed?
+        changed
```

Comment:
could this be simplified to:

```ruby
changed_levels = level_file_paths.map do |path|
  load_custom_level(path, level_index)
end.select(&:changed?)
```

(cf. https://github.com/code-dot-org/code-dot-org/compare/activerecord-import#diff-427397e290ab7c237640aabb2d0db238R4)

Reply:
:man_facepalming: Yep! Thanks for catching that.